content/blog/eng/how-i-became-a-gitlab-contributor-to-fix-an-issue-in-our-projects-cicd-pipeline.md

```diff
@@ -0,0 +1,387 @@
+
+This, coupled with the shared `variables` between `jobs` and `services`, enabled me to provide some kind of "scoped" `variables` to each **mock server** instance in the following way:
+
+```yaml
+job-1:
+    services:
+        - name: $MOCK_IMAGE:$COMMIT_ID
+          alias: company-1-mock
+          entrypoint: ["/bin/sh"]
+          command: ["-c", "echo \"$COMPANY_1_ENVIRONMENT\" > init.env && source init.env; <original Docker command>"]
+        - name: $MOCK_IMAGE:$COMMIT_ID
+          alias: company-2-mock
+          entrypoint: ["/bin/sh"]
+          command: ["-c", "echo \"$COMPANY_2_ENVIRONMENT\" > init.env && source init.env; <original Docker command>"]
+    script:
+        - npm run tests # Generic wrapper to test all the companies
+    variables:
+        COMPANY_1_ENVIRONMENT: |
+            export COMPANY_NAME=company-1
+            export OTHER_VAR=...
+        COMPANY_2_ENVIRONMENT: |
+            export COMPANY_NAME=company-2
+            export OTHER_VAR=...
+        ...
+```
+
+**Success, it works! ✅**
+
+We basically replaced the startup process for each of our `services` in order to load the `variables` from a local file, and every `service` is provided with a different file. This kind of works like the `env_file` directive in the `docker-compose.yaml` file that we've seen before.
+
+⚠️ Being this a workaround, however, there are some implications that need to be considered:
+
+* we can't define the `variables` as plain `YAML` properties as we did with **Docker Compose**
+* we still need to define a shared `variable` for each `service` at `job` level, meaning that `company-2-mock` can do `echo $COMPANY_1_ENVIRONMENT` and retrieve all the values for the other **mock server**, thus breaking the isolation between `services`
+* we need to manually add the original **Docker** `command` at the end of our overriden one, meaning that we need to keep it updated in case the base **Docker** image changes
+
+To overcome these issues we need a less "hacky" way to deal with our scenario, even if **GitLab** is not supporting it. Luckily **GitLab** is an **opensource** software, and this allows people to contribute to their code-base even if not directly employed by **GitLab** itself.
+
+As I said before I'm not a developer anymore, but I decided to go ahead and took my chance at adding the feature that we needed to improve our **CI/CD** pipelines.
+
+## Becoming a GitLab contributor to solve our issue with CI/CD
+
+>ℹ️
+>
+>To develop on **GitLab**'s platform you need the [GitLab Development Kit](https://gitlab.com/gitlab-org/gitlab-development-kit). The kit is well documented so I won't go into the details of the workflow.
+
+If you're not familiar with **GitLab**'s code-base, the initial experience will be daunting at best as their repository is quite huge. Lacking experience with **Ruby** didn't help either, but eventually I got a grip of which classes I needed to fiddle with in order to have my feature working.
```

Comment:
If you're not familiar with **GitLab**'s codebase, the initial experience will be daunting at best, as their repository is quite large. Furthermore, the lack of experience with **Ruby** didn't help, but eventually I got a grip on which classes I needed to fiddle with to make my feature work.